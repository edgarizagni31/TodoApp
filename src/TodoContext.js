import { createContext, useEffect, useReducer, useState } from 'react';
import { todoReducer } from './helpers/todoReducer';

export const TodoContext = createContext();

const  TodoProvider = ( {children} ) => {
    const init = () => JSON.parse(localStorage.getItem('todos')) || [];
    const [ todos, dispatch ] = useReducer( todoReducer, [], init );
    const [ listTodos, setListTodos ] = useState(todos);

    useEffect( 
        () => {
            localStorage.setItem('todos', JSON.stringify(todos));
            setListTodos( todos );
        }
    , [ todos ])
   
    const filterComplete =   () => {
      setListTodos( todos.filter( (t) => t.done === true ) );
    }
  
    const filterActives =  () => {
      setListTodos( todos.filter( (t) => t.done === false ) )
    }

    const resetTodos = () => {
      setListTodos( todos );
    }

    return (
        <TodoContext.Provider 
            value = {
                {
                    listTodos,
                    dispatch,
                    resetTodos,
                    filterActives,
                    filterComplete
                }
            }
        >
            {children}
        </TodoContext.Provider>
    )
}

export default TodoProvider;
