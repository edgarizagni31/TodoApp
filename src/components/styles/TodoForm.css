.form {
  margin-bottom: 20px;
}

.form__input {
  background-color: hsl(235, 24%, 19%);
  border: none;
  border-radius: 4px;
  color: #ffffff;
  font-size: 18px;
  font-weight: 400;
  height: 38px;
  outline: 0;
  margin-bottom: 20px;
  padding-left: 10px;
  width: 100%;
}

.form__btn {
  height: 32px;
  width: 100%;
  outline: none;
  font-size: 14px;
  background-color: hsl(234,39%,85%);
  border: none;
}
