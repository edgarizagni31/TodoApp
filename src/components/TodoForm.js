import { useContext, useRef } from 'react';
import { TodoContext } from '../TodoContext';

import './styles/TodoForm.css';

const TodoForm = () => {
    const inputEl = useRef(null);
    const { dispatch } = useContext(TodoContext);
    
    const handleSubmit = (e) => {
        e.preventDefault();
        let todo;

        if (inputEl.current.value !== '') {
            todo = {
                id: Date.now(),
                desc: inputEl.current.value,
                done: false
            };

            dispatch( {  type: 'add', payload: todo } );
            inputEl.current.value = '';
        }else {
            alert('Este campo no puede estar vacio');
        }
    }

    return (
        <form className = 'form' onSubmit = {handleSubmit}>
            <input
                type = 'text'
                name = 'desc'
                placeholder = 'Crear un nuevo todo...'
                ref = {inputEl}
                autoComplete = 'off'
                className = 'form__input'
            />
        </form>
    )
}

export default TodoForm;

