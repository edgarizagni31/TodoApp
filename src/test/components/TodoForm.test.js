import  { TodoContext } from '../../TodoContext';
import  { mount } from 'enzyme';
import TodoForm from '../../components/TodoForm';

describe('test to TodoForm', () => {
    const dispatch = jest.fn();

    const wrapper = mount(
        <TodoContext.Provider value = {{  dispatch }}>
            <TodoForm />
        </TodoContext.Provider>
    );

    test('should be displayed correctly', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('should t call dispatch', () => {
        window.alert = jest.fn();

        wrapper.find('form').prop('onSubmit')( { preventDefault(){} } );
        expect(dispatch).not.toHaveBeenCalled();
        expect( alert ).toBeCalledWith('Este campo no puede estar vacio');
    });

    test('should call dispatch', () => {

        wrapper.find('input').simulate('change', { target: { value: 'hola' } });
        wrapper.find('form').prop('onSubmit')( { preventDefault(){} } );
        expect(dispatch).toHaveBeenCalledTimes(1);
    });

})
