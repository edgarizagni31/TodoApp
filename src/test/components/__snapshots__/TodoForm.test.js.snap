// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test to TodoForm should be displayed correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    className="form__input"
    name="desc"
    onChange={[Function]}
    placeholder="Crear un nuevo todo..."
    type="text"
  />
</form>
`;
